### ISSUE TITLE
XCM - CurrencyAdapter doesn't check result

### ISSUE THREAD

**Issue by Dinonard** [paritytech/polkadot#5823]
Hi!

I've noticed that in the [currency_adapter](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-builder/src/currency_adapter.rs#L145) we don't check whether calling `deposit_creating` was successful or not.

A colleague of mine was sending DOTs back to Polkadot and although XCM execution was labeled as successful, he didn't receive any DOTs. His account didn't satisfy the ED requirement and the amount being sent was only `0.1` DOTs so it wasn't sufficient to create a new account.

It was quite misleading since we'd expect the execution to fail with some error.

Is this by design? Or should it be changed to result in an error?

**Comment by bkchr** [comment]
CC @KiChjang

**Comment by KiChjang** [comment]
`deposit_creating` doesn't return a `Result`, instead it returns a `0` amount on error, but it can mean a lot of things, not simply just because an amount is less than ED.

This doesn't really sound like an XCM issue but an issue with the `Currency` trait.

**Comment by Dinonard** [comment]
You're right, `0` is a valid amount since I could repeat my scenario by sending exactly the amount needed to cover the execution fee.

I guess adding checks to the amount in the holding register & comparing it to the imbalance value is over-complicating it?

I could also ask in `substrate` about changing the `Currency` trait (a bit unlikely :slightly_smiling_face:) or maybe to add another trait that would have the same methods but would return `Resut<...>` instead (would be backwards compatible).
